File: app/models/cmd/sponsorship.go
package cmd

import (
	"time"

	"github.com/Spicy-Bush/fider-tarkov-community/app/models/entity"
)

type CreateSponsorshipPackage struct {
	Slug         string
	Name         string
	Description  string
	Slots        string
	DurationDays int
	Sort         int
	Result       *entity.SponsorshipPackage
}

type UpdateSponsorshipPackage struct {
	ID           int
	Slug         string
	Name         string
	Description  string
	Slots        string
	DurationDays int
	Sort         int
	Result       *entity.SponsorshipPackage
}

type DeleteSponsorshipPackage struct {
	ID int
}

type CreateSponsorshipCampaign struct {
	Name      string
	Advertiser string
	StartAt   time.Time
	EndAt     time.Time
	Weight    int
	Locale    string
	Enabled   bool
	PackageID *int
	Result    *entity.SponsorshipCampaign
}

type UpdateSponsorshipCampaign struct {
	ID            int
	Name          string
	Advertiser    string
	StartAt       time.Time
	EndAt         time.Time
	Weight        int
	Locale        string
	Enabled       bool
	PackageID     *int
	ConfigVersion int // required OCC token; store rejects mismatch and bumps
	Result        *entity.SponsorshipCampaign
}

type DeleteSponsorshipCampaign struct {
	ID int
}

type IncrementSponsorshipClick struct {
	ID int
}

File: public/services/actions/sponsorship.ts
import { http, Result } from "@fider/services/http"
import { PublicSponsorshipCampaign, SponsorshipCampaign, SponsorshipPackage } from "@fider/models"

export const listSponsorshipPackages = (): Promise<Result<SponsorshipPackage[]>> => {
  return http.get<SponsorshipPackage[]>("/api/v1/sponsorship/packages")
}

export const createSponsorshipPackage = (body: {
  slug: string
  name: string
  description: string
  slots: string
  durationDays: number
  sort: number
}): Promise<Result<SponsorshipPackage>> => {
  return http.post<SponsorshipPackage>("/api/v1/sponsorship/packages", body)
}

export const updateSponsorshipPackage = (
  id: number,
  body: {
    slug: string
    name: string
    description: string
    slots: string
    durationDays: number
    sort: number
  }
): Promise<Result<SponsorshipPackage>> => {
  return http.put<SponsorshipPackage>(`/api/v1/sponsorship/packages/${id}`, body)
}

export const deleteSponsorshipPackage = (id: number): Promise<Result> => {
  return http.delete(`/api/v1/sponsorship/packages/${id}`)
}

export const listSponsorshipCampaigns = (): Promise<Result<SponsorshipCampaign[]>> => {
  return http.get<SponsorshipCampaign[]>("/api/v1/sponsorship/campaigns")
}

export const createSponsorshipCampaign = (body: Partial<SponsorshipCampaign> & {
  name: string
  slotId: string
  clickUrl: string
  startAt: string
  endAt: string
}): Promise<Result<SponsorshipCampaign>> => {
  return http.post<SponsorshipCampaign>("/api/v1/sponsorship/campaigns", body)
}

export const updateSponsorshipCampaign = (
  id: number,
  body: Partial<SponsorshipCampaign> & {
    name: string
    slotId: string
    clickUrl: string
    startAt: string
    endAt: string
  }
): Promise<Result<SponsorshipCampaign>> => {
  return http.put<SponsorshipCampaign>(`/api/v1/sponsorship/campaigns/${id}`, body)
}

export const deleteSponsorshipCampaign = (id: number): Promise<Result> => {
  return http.delete(`/api/v1/sponsorship/campaigns/${id}`)
}

export const getActiveSponsorship = (
  slot: string,
  locale: string
): Promise<Result<PublicSponsorshipCampaign | Record<string, never>>> => {
  const q = new URLSearchParams({ slot, locale })
  return http.get(`/api/v1/sponsorship/active?${q.toString()}`)
}
